runtime/complex.go

```diff
@@ -130,3 +134,130 @@ func complexCoerce(o *Object) (complex128, bool) {
 	}
 	return complex(floatO, 0.0), true
 }
+
+func complexNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	if argc == 0 {
+		return newObject(t), nil
+	}
+	if argc > 2 {
+		return nil, f.RaiseType(TypeErrorType, "'__new__' of 'complex' requires at most 2 arguments")
+	}
+	if t != ComplexType {
+		// Allocate a plain complex then copy it's value into an object
+		// of the complex subtype.
+		x, raised := complexNew(f, ComplexType, args, nil)
+		if raised != nil {
+			return nil, raised
+		}
+		result := toComplexUnsafe(newObject(t))
+		result.value = toComplexUnsafe(x).Value()
+		return result.ToObject(), nil
+	}
+	o := args[0]
+	if argc == 1 {
+		if complexSlot := o.typ.slots.Complex; complexSlot != nil {
+			result, raised := complexSlot.Fn(f, o)
+			if raised != nil {
+				return nil, raised
+			}
+			if !result.isInstance(ComplexType) {
+				format := "__complex__ returned non-complex (type %s)"
+				return nil, f.RaiseType(TypeErrorType, fmt.Sprintf(format, result.typ.Name()))
+			}
+			return result, nil
+		}
+		if floatSlot := o.typ.slots.Float; floatSlot != nil {
+			result, raised := floatConvert(floatSlot, f, o)
+			if raised != nil {
+				return nil, raised
+			}
+			f := toFloatUnsafe(result).Value()
```

Comment:
If floatConvert() returned (*Float, *BaseException) then this unsafe cast would be unnecessary.